File: app/src/main/java/com/dss/sdatabase/dao/BaseTable.java
package com.dss.sdatabase.dao;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;

import com.dss.sdatabase.annotations.BaseDBFieldName;
import com.dss.sdatabase.annotations.BaseDBMethodSetName;
import com.dss.sdatabase.annotations.BaseDBPrimaryKey;
import com.dss.sdatabase.annotations.BaseDBType;
import com.dss.sdatabase.constant.ConstantException;
import com.dss.sdatabase.exceptions.InvalidTypeException;
import com.dss.sdatabase.model.BDCreate;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by gustavo.vieira on 04/05/2015.
 */
public class BaseTable {

    public BaseDB baseDB;
    private SQLiteDatabase db = null;
    private Context context;
    private String DBname;
    private int DBversion;

    private String tableName;
    private ArrayList<String> fields;
    private ArrayList<String> types;
    private String pk;


    protected BaseTable(Context context, Class modelClass,String DBname, int DBversion){
        super();
        this.DBname = DBname;
        this.DBversion = DBversion;
        baseDB = new BaseDB(context, DBname, DBversion);
        this.context = context;
        Field[] f = modelClass.getDeclaredFields();

        tableName = modelClass.getSimpleName();

        fields = new ArrayList<>();
        types = new ArrayList<>();

        for(int i = 0; i < f.length; i++){
            f[i].setAccessible(true);
            if (f[i].isAnnotationPresent(BaseDBFieldName.class)){
                fields.add(f[i].getAnnotation(BaseDBFieldName.class).value());
                types.add(f[i].getAnnotation(BaseDBType.class).value());
            }
            if(f[i].isAnnotationPresent(BaseDBPrimaryKey.class)){
                pk = f[i].getAnnotation(BaseDBFieldName.class).value();
            }
        }

    }


    protected void createTable(){
        ArrayList<BDCreate> bdCreates = new ArrayList<>();
        BDCreate bdCreate;

        for(int i = 0; i < fields.size(); i++) {
            bdCreate = new BDCreate();
            bdCreate.setFieldName(fields.get(i));
            bdCreate.setFieldType(types.get(i));

            bdCreates.add(bdCreate);

        }

        openCoonection();
        baseDB.createTable(db, tableName, bdCreates, pk);
        closeConnection();
    }

    private void openCoonection(){
        this.baseDB = new BaseDB(context, DBname, DBversion);
        baseDB.setTable(tableName);
        db = baseDB.getWritableDatabase();
    }

    private void closeConnection() {
        try {
            if (null != db) {
                db.close();
                db = null;
                baseDB.close();
            }

        } catch (Exception e) {
            Log.e("Erro:", "Erro ao fechar conexoes");
        }
    }


    protected void insert(ArrayList<Object> receiveObject){

        openCoonection();
        Object insertObject;
        for(Object object : receiveObject) {
            insertObject = object;
            try {
                baseDB.insert(tableName, insertObject);
            } catch (InvalidTypeException invalidTypeException) {
                invalidTypeException.printStackTrace();
            }
        }
        closeConnection();

    }

    protected ArrayList<Object> selectList(Class aClass)
            throws ClassNotFoundException, NoSuchMethodException,
            IllegalAccessException, InvocationTargetException,
            InstantiationException, InvalidTypeException {

        Class<?> clazz = Class.forName(aClass.getName());
        Constructor<?> ctor = clazz.getConstructor();
        ctor.setAccessible(true);

        ArrayList<Object> list = new ArrayList<>();
        Object object;


        openCoonection();

        Cursor c = baseDB.get(tableName,fields);

        try {
            while (c.moveToNext()) {
                object = ctor.newInstance();

                Method[] methods = aClass.getMethods();

                for(int i = 0;i < methods.length;i++) {
                    methods[i].setAccessible(true);


                    if (methods[i].isAnnotationPresent(BaseDBMethodSetName.class)) {

                            Type parameterizedType = (Type) methods[i].getGenericParameterTypes()[0];

                        if (parameterizedType == Integer.class||
                                parameterizedType == int.class) {

                            methods[i].invoke(object,c.getInt(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));


                        } else if (parameterizedType == Long.class) {

                            methods[i].invoke(object,c.getLong(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                        }else if (parameterizedType == Double.class) {

                            methods[i].invoke(object,c.getDouble(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                            //Verify if variable is text
                        }else if (parameterizedType == String.class ||
                                parameterizedType == char.class) {

                            methods[i].invoke(object,c.getString(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                        }else{
                            throw new InvalidTypeException(ConstantException.getINVALIDTYPEEXCEPTION());
                        }
                    }
                }

                list.add(object);

            }
        } finally {
            c.close();
        }
        return list;

    }

    protected Object selectWhere(Class aClass,ArrayList<String> field,String[] values)
            throws ClassNotFoundException, NoSuchMethodException,
            IllegalAccessException, InvocationTargetException,
            InstantiationException, InvalidTypeException{


        Class<?> clazz = Class.forName(aClass.getName());
        Constructor<?> ctor = clazz.getConstructor();
        ctor.setAccessible(true);

        Object object = ctor.newInstance();

        Cursor c = baseDB.getWhere(tableName, fields, field, values);

        try {
            if (c.moveToNext()) {
                Method[] methods = aClass.getMethods();

                for(int i = 0;i < methods.length;i++) {
                    methods[i].setAccessible(true);

                    if (methods[i].isAnnotationPresent(BaseDBMethodSetName.class)) {

                        Type parameterizedType = (Type) methods[i].getGenericParameterTypes()[0];

                        if (parameterizedType == Integer.class||
                                parameterizedType == int.class) {

                            methods[i].invoke(object,c.getInt(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));


                        } else if (parameterizedType == Long.class) {

                            methods[i].invoke(object,c.getLong(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                        }else if (parameterizedType == Double.class) {

                            methods[i].invoke(object,c.getDouble(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                            //Verify if variable is text
                        }else if (parameterizedType == String.class ||
                                parameterizedType == char.class) {

                            methods[i].invoke(object,c.getString(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                        }else{
                            throw new InvalidTypeException(ConstantException.getINVALIDTYPEEXCEPTION());
                        }
                    }
                }
            }
        } finally {
            c.close();
        }

        return object;
    }

    protected ArrayList<Object> selectListWhere(Class aClass,ArrayList<String>field,String[] values)
            throws ClassNotFoundException, NoSuchMethodException,
            IllegalAccessException, InvocationTargetException,
            InstantiationException, InvalidTypeException{

        Class<?> clazz = Class.forName(aClass.getName());
        Constructor<?> ctor = clazz.getConstructor();
        ctor.setAccessible(true);

        ArrayList<Object> list = new ArrayList<>();
        Object object;

        openCoonection();
        Cursor c = baseDB.getWhere(tableName, fields, field, values);

        try {
            while (c.moveToNext()) {
                object = ctor.newInstance();

                Method[] methods = aClass.getMethods();

                for(int i = 0;i < methods.length;i++) {
                    methods[i].setAccessible(true);


                    if (methods[i].isAnnotationPresent(BaseDBMethodSetName.class)) {

                        Type parameterizedType = (Type) methods[i].getGenericParameterTypes()[0];

                        if (parameterizedType == Integer.class||
                                parameterizedType == int.class) {

                            methods[i].invoke(object,c.getInt(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));


                        } else if (parameterizedType == Long.class) {

                            methods[i].invoke(object,c.getLong(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                        }else if (parameterizedType == Double.class) {

                            methods[i].invoke(object,c.getDouble(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                            //Verify if variable is text
                        }else if (parameterizedType == String.class ||
                                parameterizedType == char.class) {

                            methods[i].invoke(object,c.getString(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                        }else{
                            throw new InvalidTypeException(ConstantException.getINVALIDTYPEEXCEPTION());
                        }
                    }
                }

                list.add(object);

            }
        } finally {
            c.close();
        }
        return list;
    }


    protected ArrayList<Object> selectRawQuery(Class aClass, String query, String[] values)
            throws ClassNotFoundException, NoSuchMethodException,
            IllegalAccessException, InvocationTargetException,
            InstantiationException, InvalidTypeException{


        Class<?> clazz = Class.forName(aClass.getName());
        Constructor<?> ctor = clazz.getConstructor();
        ctor.setAccessible(true);

        ArrayList<Object> list = new ArrayList<>();

        Object object = ctor.newInstance();

        Cursor c = baseDB.getRawQuery(query,values);

        try {
            while (c.moveToNext()) {
                object = ctor.newInstance();

                Method[] methods = aClass.getMethods();

                for(int i = 0;i < methods.length;i++) {
                    methods[i].setAccessible(true);

                    if (methods[i].isAnnotationPresent(BaseDBMethodSetName.class)) {

                        Type parameterizedType = (Type) methods[i].getGenericParameterTypes()[0];

                        if (parameterizedType == Integer.class||
                                parameterizedType == int.class) {

                            methods[i].invoke(object,c.getInt(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));


                        } else if (parameterizedType == Long.class) {

                            methods[i].invoke(object,c.getLong(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                        }else if (parameterizedType == Double.class) {

                            methods[i].invoke(object,c.getDouble(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                            //Verify if variable is text
                        }else if (parameterizedType == String.class ||
                                parameterizedType == char.class) {

                            methods[i].invoke(object,c.getString(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                        }else{
                            throw new InvalidTypeException(ConstantException.getINVALIDTYPEEXCEPTION());
                        }
                    }
                }
                list.add(object);

            }
        } finally {
            c.close();
        }

        return list;
    }

    protected ArrayList<Object> selectListRawQuery(Class aClass,String query,String[] values)
            throws ClassNotFoundException, NoSuchMethodException,
            IllegalAccessException, InvocationTargetException,
            InstantiationException, InvalidTypeException{

        Class<?> clazz = Class.forName(aClass.getName());
        Constructor<?> ctor = clazz.getConstructor();
        ctor.setAccessible(true);

        ArrayList<Object> list = new ArrayList<>();
        Object object;

        openCoonection();
        Cursor c = baseDB.getRawQuery(query,values);

        try {
            while (c.moveToNext()) {
                object = ctor.newInstance();

                Method[] methods = aClass.getMethods();

                for(int i = 0;i < methods.length;i++) {
                    methods[i].setAccessible(true);


                    if (methods[i].isAnnotationPresent(BaseDBMethodSetName.class)) {

                        Type parameterizedType = (Type) methods[i].getGenericParameterTypes()[0];

                        if (parameterizedType == Integer.class||
                                parameterizedType == int.class) {

                            methods[i].invoke(object,c.getInt(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));


                        } else if (parameterizedType == Long.class) {

                            methods[i].invoke(object,c.getLong(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                        }else if (parameterizedType == Double.class) {

                            methods[i].invoke(object,c.getDouble(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                            //Verify if variable is text
                        }else if (parameterizedType == String.class ||
                                parameterizedType == char.class) {

                            methods[i].invoke(object,c.getString(c.getColumnIndex(methods[i].getAnnotation(BaseDBMethodSetName.class).value())));

                        }else{
                            throw new InvalidTypeException(ConstantException.getINVALIDTYPEEXCEPTION());
                        }
                    }
                }

                list.add(object);

            }
        } finally {
            c.close();
        }
        return list;
    }

}

File: app/src/main/java/com/dss/sdatabase/dao/BaseTable.kt
package com.dss.sdatabase.dao

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.util.Log
import com.dss.sdatabase.annotations.BaseDBFieldName
import com.dss.sdatabase.annotations.BaseDBPrimaryKey
import com.dss.sdatabase.annotations.BaseDBType
import com.dss.sdatabase.constant.ConstantException
import com.dss.sdatabase.exceptions.InvalidTypeException
import com.dss.sdatabase.model.BDCreate
import java.lang.reflect.InvocationTargetException
import java.util.*

/**
 * Created by gustavo.vieira on 04/05/2015.
 */
open class BaseTable protected constructor(val context: Context, modelClass: Class<*>, val DBname: String, val DBversion: Int) {

    var baseDB: BaseDB
    private var db: SQLiteDatabase? = null
    private var aClass: Class<*>? = null
    private val tableName: String
    private val fields: ArrayList<String>
    private val types: ArrayList<String>
    private var pk: String? = null

    companion object {
        init {

        }
    }

    init {
        baseDB = BaseDB(context, DBname, DBversion)
        val f = modelClass.declaredFields
        aClass = modelClass

        tableName = modelClass.simpleName

        fields = ArrayList<String>()
        types = ArrayList<String>()

        for (i in f.indices) {
            f[i].isAccessible = true
            if (f[i].isAnnotationPresent(BaseDBFieldName::class.java)) {
                fields.add(f[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)
                types.add(f[i].getAnnotation<BaseDBType>(BaseDBType::class.java).value)
            }
            if (f[i].isAnnotationPresent(BaseDBPrimaryKey::class.java)) {
                pk = f[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value
            }
        }

    }


    protected fun createTable() {
        val bdCreates = ArrayList<BDCreate>()
        var bdCreate: BDCreate

        for (i in fields.indices) {
            bdCreate = BDCreate()
            bdCreate.fieldName = fields[i]
            bdCreate.fieldType = types[i]

            bdCreates.add(bdCreate)

        }

        openCoonection()
        baseDB.createTable(db!!, tableName, bdCreates, pk)
        closeConnection()
    }

    private fun openCoonection() {
        this.baseDB = BaseDB(context, DBname, DBversion)
        baseDB.table = tableName
        db = baseDB.writableDatabase
    }

    private fun closeConnection() {
        try {
            if (null != db) {
                db!!.close()
                db = null
                baseDB.close()
            }

        } catch (e: Exception) {
            Log.e("Erro:", "Erro ao fechar conexoes")
        }

    }


    protected fun insert(receiveObject: ArrayList<Any>) {

        openCoonection()
        var insertObject: Any
        for (`object` in receiveObject) {
            insertObject = `object`
            try {
                baseDB.insert(tableName, insertObject)
            } catch (invalidTypeException: InvalidTypeException) {
                invalidTypeException.printStackTrace()
            }

        }
        closeConnection()

    }

    @Throws(ClassNotFoundException::class, IllegalAccessException::class, InvocationTargetException::class, InstantiationException::class, InvalidTypeException::class)
    protected fun selectList(): ArrayList<Any> {

        val clazz = Class.forName(aClass!!.name)
        val ctor = clazz.getConstructor()
        ctor.isAccessible = true

        val list = ArrayList<Any>()
        var `object`: Any


        openCoonection()

        val c = baseDB.get(tableName, fields)

        try {
            while (c.moveToNext()) {
                `object` = ctor.newInstance()

                val variables = aClass!!.fields

                for (i in variables.indices) {
                    variables[i].isAccessible = true


                    if (variables[i].isAnnotationPresent(BaseDBFieldName::class.java)) {

                        val parameterizedType = variables[i].genericType

                        if (parameterizedType === Int::class.java || parameterizedType === Int::class.javaPrimitiveType) {

                            variables[i].set(`object`, c.getInt(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))


                        } else if (parameterizedType === Long::class.java) {

                            variables[i].set(`object`, c.getLong(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                        } else if (parameterizedType === Double::class.java) {

                            variables[i].set(`object`, c.getDouble(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                            //Verify if variable is text
                        } else if (parameterizedType === String::class.java || parameterizedType === Char::class.javaPrimitiveType) {

                            variables[i].set(`object`, c.getString(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                        } else {
                            throw InvalidTypeException(ConstantException.invalidtypeexception)
                        }
                    }
                }

                list.add(`object`)

            }
        } finally {
            c.close()
        }
        return list

    }

    @Throws(ClassNotFoundException::class, IllegalAccessException::class, InvocationTargetException::class, InstantiationException::class, InvalidTypeException::class)
    protected fun selectWhere(field: ArrayList<String>, values: Array<String>): Any {


        val clazz = Class.forName(aClass!!.name)
        val ctor = clazz.getConstructor()
        ctor.isAccessible = true

        val `object` = ctor.newInstance()

        val c = baseDB.getWhere(tableName, fields, field, values)

        try {
            if (c.moveToNext()) {
                val variables = aClass!!.fields

                for (i in variables.indices) {
                    variables[i].isAccessible = true

                    if (variables[i].isAnnotationPresent(BaseDBFieldName::class.java)) {
                        var parameterizedType = variables[i].genericType

                        if (parameterizedType === Int::class.java || parameterizedType === Int::class.javaPrimitiveType) {

                            variables[i].set(`object`, c.getInt(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                        } else if (parameterizedType === Long::class.java) {

                            variables[i].set(`object`, c.getLong(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                        } else if (parameterizedType === Double::class.java) {

                            variables[i].set(`object`, c.getDouble(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                            //Verify if variable is text
                        } else if (parameterizedType === String::class.java || parameterizedType === Char::class.javaPrimitiveType) {

                            variables[i].set(`object`, c.getString(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                        } else {
                            throw InvalidTypeException(ConstantException.invalidtypeexception)
                        }
                    }
                }
            }
        } finally {
            c.close()
        }

        return `object`
    }

    @Throws(ClassNotFoundException::class, IllegalAccessException::class, InvocationTargetException::class, InstantiationException::class, InvalidTypeException::class)
    protected fun selectListWhere(field: ArrayList<String>, values: Array<String>): ArrayList<Any> {

        val clazz = Class.forName(aClass!!.name)
        val ctor = clazz.getConstructor()
        ctor.isAccessible = true

        val list = ArrayList<Any>()
        var `object`: Any

        openCoonection()
        val c = baseDB.getWhere(tableName, fields, field, values)

        try {
            while (c.moveToNext()) {
                `object` = ctor.newInstance()

                val variables = aClass!!.fields

                for (i in variables.indices) {
                    variables[i].isAccessible = true


                    if (variables[i].isAnnotationPresent(BaseDBFieldName::class.java)) {

                        val parameterizedType = variables[i].genericType

                        if (parameterizedType === Int::class.java || parameterizedType === Int::class.javaPrimitiveType) {

                            variables[i].set(`object`, c.getInt(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))


                        } else if (parameterizedType === Long::class.java) {

                            variables[i].set(`object`, c.getLong(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                        } else if (parameterizedType === Double::class.java) {

                            variables[i].set(`object`, c.getDouble(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                            //Verify if variable is text
                        } else if (parameterizedType === String::class.java || parameterizedType === Char::class.javaPrimitiveType) {

                            variables[i].set(`object`, c.getString(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                        } else {
                            throw InvalidTypeException(ConstantException.invalidtypeexception)
                        }
                    }
                }

                list.add(`object`)

            }
        } finally {
            c.close()
        }
        return list
    }


    @Throws(ClassNotFoundException::class, IllegalAccessException::class, InvocationTargetException::class, InstantiationException::class, InvalidTypeException::class)
    protected fun selectRawQuery(query: String, values: Array<String>): ArrayList<Any> {


        val clazz = Class.forName(aClass!!.name)
        val ctor = clazz.getConstructor()
        ctor.isAccessible = true

        val list = ArrayList<Any>()

        var `object`: Any = ctor.newInstance()

        val c = baseDB.getRawQuery(query, values)

        try {
            while (c.moveToNext()) {
                `object` = ctor.newInstance()

                val variables = aClass!!.fields

                for (i in variables.indices) {
                    variables[i].isAccessible = true

                    if (variables[i].isAnnotationPresent(BaseDBFieldName::class.java)) {

                        val parameterizedType = variables[i].genericType

                        if (parameterizedType === Int::class.java || parameterizedType === Int::class.javaPrimitiveType) {

                            variables[i].set(`object`, c.getInt(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))


                        } else if (parameterizedType === Long::class.java) {

                            variables[i].set(`object`, c.getLong(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                        } else if (parameterizedType === Double::class.java) {

                            variables[i].set(`object`, c.getDouble(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                            //Verify if variable is text
                        } else if (parameterizedType === String::class.java || parameterizedType === Char::class.javaPrimitiveType) {

                            variables[i].set(`object`, c.getString(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                        } else {
                            throw InvalidTypeException(ConstantException.invalidtypeexception)
                        }
                    }
                }
                list.add(`object`)

            }
        } finally {
            c.close()
        }

        return list
    }

    @Throws(ClassNotFoundException::class, IllegalAccessException::class, InvocationTargetException::class, InstantiationException::class, InvalidTypeException::class)
    protected fun selectListRawQuery(query: String, values: Array<String>): ArrayList<Any> {

        val clazz = Class.forName(aClass!!.name)
        val ctor = clazz.getConstructor()
        ctor.isAccessible = true

        val list = ArrayList<Any>()
        var `object`: Any

        openCoonection()
        val c = baseDB.getRawQuery(query, values)

        try {
            while (c.moveToNext()) {
                `object` = ctor.newInstance()

                val variables = aClass!!.fields

                for (i in variables.indices) {
                    variables[i].isAccessible = true


                    if (variables[i].isAnnotationPresent(BaseDBFieldName::class.java)) {

                        val parameterizedType = variables[i].genericType

                        if (parameterizedType === Int::class.java || parameterizedType === Int::class.javaPrimitiveType) {

                            variables[i].set(`object`, c.getInt(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))


                        } else if (parameterizedType === Long::class.java) {

                            variables[i].set(`object`, c.getLong(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                        } else if (parameterizedType === Double::class.java) {

                            variables[i].set(`object`, c.getDouble(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                            //Verify if variable is text
                        } else if (parameterizedType === String::class.java || parameterizedType === Char::class.javaPrimitiveType) {

                            variables[i].set(`object`, c.getString(c.getColumnIndex(variables[i].getAnnotation<BaseDBFieldName>(BaseDBFieldName::class.java).value)))

                        } else {
                            throw InvalidTypeException(ConstantException.invalidtypeexception)
                        }
                    }
                }

                list.add(`object`)

            }
        } finally {
            c.close()
        }
        return list
    }


    protected fun deleteAll(): Boolean {

        val b = baseDB.DeleteFieldsTable(tableName)

        return b!!
    }

    protected fun deleteAll(field: String, values: Array<String>): Boolean {

        val b = baseDB.DeleteFieldsTableWhere(tableName, field, values)

        return b!!
    }


    protected fun dropTable() {
        baseDB.dropTable(db!!, tableName)
    }

}
